File: write.py
#!/usr/bin/python3
import socket
import sys
from writerGUI import Ui_MainWindow
from PyQt5.QtWidgets import QApplication, QDialog, QMainWindow, QShortcut
from PyQt5.QtGui import QKeySequence
from PyQt5.QtCore import Qt
import time
import re
import wave
import linecache
import sys

def PrintException():
    exc_type, exc_obj, tb = sys.exc_info()
    f = tb.tb_frame
    lineno = tb.tb_lineno
    filename = f.f_code.co_filename
    linecache.checkcache(filename)
    line = linecache.getline(filename, lineno, f.f_globals)
    print ('EXCEPTION IN ({}, LINE {} "{}"): {}'.format(filename, lineno, line.strip(), exc_obj))


readerIP = "192.168.240.131"
readerPort = 100
logDirectory = 'logs'

def writeData(data, readerIP, readerPort):
	print("WRITE DATA")
	try:
		s = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.SOL_TCP)
		s.connect((readerIP, readerPort))
	except:
		raise Exception('NetworkError: Socket creation failed.')

	for i in range(12-len(data)):
		data = data + chr(0)
	for j in range(6):
		b2 = (data[j*2])
		b1 = (data[j*2+1])

		# Sending Write request... 10,255,10,137,0,0,0,0,1, 7 - position to write in, byte1, byte2, CheckDigitValue
		print("Writing 2 bytes")
		cmd = bytearray([10, 255, 10, 137, 0, 0, 0, 0, 1, 7 - j, ord(b1) ,ord(b2), CheckDigit(7-j, ord(b1), ord(b2))])
		s.send(cmd)

		# Reading response...
		out = s.recv(2048)
		print("Response: " + " ".join("%02x" % b for b in out))
		if out[3] == 82:
			raise Exception('Write Error: Check if item placed on the writer.')	#happened when no tag was there on the device.
	
	return readData(readerIP, readerPort)


def readData(readerIP, readerPort):
	try:
		s = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.SOL_TCP)
		s.connect((readerIP, readerPort))
	except:
		raise Exception('NetworkError: Socket creation failed.')

	print("Sending Read request.")
	cmd = bytearray([10, 255, 2, 128, 117])
	s.send(cmd)

	# Reading response
	out = s.recv(2048)
	cnt = out[5]
	print("Response: " + " ".join("%02x" % b for b in out))

	print("Sending get tag data request.")
	cmd = bytearray([10, 255, 3, 65, 16, 163])
	s.send(cmd)

	# Reading response
	out = s.recv(2048)
	print("Response: " + " ".join("%02x" % b for b in out))
	if out[4] > 1:
		raise Exception("WARNING: More than one tags in range!!!")
	elif out[4] == 0:
		raise Exception("WARNING: No tags in range!!!")
	out = out[7:7+12][::-1]
	if out[1] == 0x9e:
		raise Exception("WARNING: Attempted to read empty tag.")
	out = out.decode()
	out = ''.join([c if ord(c) != 0 else '' for c in out])

	
	return out
	

def CheckDigit(a, b, c):
	i = a + b + c + 413
	
	if i < 255:
		i = 256 - i
	elif i < 511:
		i = 512 - i
	elif i < 1023:
		i = 1024 - i
	
	if i > 255:
		i = i - 256
	
	return i

class MainWindow(Ui_MainWindow):
	def __init__(self):
		global logDirectory
		super().__init__()
		self.window = QMainWindow()
		self.setupUi(self.window)
		self.window.show()
		self.window.move(300,300)

		self.pushButton.setDefault(True)
		self.pushButton.setAutoDefault(True)
		self.pushButton.clicked.connect(self.writeButton_callback)
		self.enterShortcut = QShortcut(QKeySequence(Qt.Key_Return), self.window)
		self.enterShortcut.activated.connect(self.writeButton_callback)

		self.readButton.clicked.connect(self.readButton_callback)
		self.pushButton_2.clicked.connect(self.patronWriteButton_callback)
		self.dataList = []
		try:
			self.count = 0
			with open(logDirectory + "/TagLog_" + time.strftime("%d%m%y") + ".csv") as fp:
				for line in fp:
					if re.search(r'\d{2}/\d{2}/\d{2}', line):  
						self.count += 1
						self.dataList.append(line.split(',')[0])
		except FileNotFoundError:
			self.textBrowser.setPlainText("Message: STARTING A NEW LOG")

		self.lcdNumber.display(self.count)
		self.lcdNumber.setStyleSheet("* { color: darkblue; background-color: black; }")

		
		
	def patronWriteButton_callback(self):
		global readerPort, readerIP, logDirectory
		try:
			data = chr(1) + self.lineEdit.text()
			ret = writeData(data, readerIP, readerPort)	
			if ret == data:
				output = "\nSUCCESSFULLY WORTE " + ret
				self.textBrowser.setPlainText(output)
			else:
				output = "WRITE ERROR! Debug: Wrote " + ret
				self.textBrowser.setPlainText('<p style="color:red;">' + output + "</p>")
		except Exception as ex:
			output = "Internal Exception: " + str(ex)
			self.textBrowser.setPlainText('<p style="color:red;"> ' + output + ' </p>')
			PrintException()	
		self.lineEdit.setFocus()
		self.lineEdit.clear()


	def readButton_callback(self):
		try:
			output = readData(readerIP, readerPort)
			self.textBrowser.setPlainText(output)
		except Exception as ex:
			output = "Internal Exception: " + str(ex)
			self.textBrowser.setPlainText('<p style="color:red;">' + output + '</p>')
			with open(logDirectory + "/ErrorLog.csv" + time.strftime("%d%m%y"), "a+") as fp:
				fp.write(output + "," + time.strftime("%d/%m/%y %H:%M:%S") + '\n')
			PrintException()	

	def writeButton_callback(self):
		global readerPort, readerIP, logDirectory
		try:
			data = self.lineEdit.text()
			if re.search(r'$\d{13}|\d{10}$', data):		#data entered is ISBN
				self.textBrowser.append("ISBN: " + data)
				with open(logDirectory + "/TagLog_" + time.strftime("%d%m%y") + ".csv", "a+") as fp:
					fp.write(data + "\n")
			else:
				data = self.removeZero(data)
				if not re.search(r'^[A-Z]{0,2}\d{1,5}$',data):
					self.textBrowser.append('<p style="color:r76778ed;">INVALID INPUT</p>')
				else:
					ret = writeData(data, readerIP, readerPort)	
					if ret == data:

						output = "\nSUCCESSFULLY WORTE " + ret
						self.textBrowser.setPlainText(output)

						if ret not in self.dataList:
							self.dataList.append(ret)
							self.count += 1
							with open(logDirectory + "/TagLog_" + time.strftime("%d%m%y") + ".csv", "a+") as fp:
								fp.write(data + "," + time.strftime("%d/%m/%y %H:%M:%S") + ',')
							self.lcdNumber.display(self.count)
						
					else:
						output = "WRITE ERROR! Debug: Wrote " + ret
						self.textBrowser.setPlainText('<p style="color:red;">' + output + "</p>")
						with open(logDirectory + "/ErrorLog.csv" + time.strftime("%d%m%y"), "a+") as fp:
							fp.write(output + "," + time.strftime("%d/%m/%y %H:%M:%S") + '\n')
		except Exception as ex:
			output = "Internal Exception: " + str(ex)
			self.textBrowser.setPlainText('<p style="color:red;"> ' + output + ' </p>')
			with open(logDirectory + "/ErrorLog.csv" + time.strftime("%d%m%y"), "a+") as fp:
				fp.write(output + "," + time.strftime("%d/%m/%y %H:%M:%S") + '\n')
			PrintException()	
		self.lineEdit.setFocus()
		self.lineEdit.clear()
				
	def removeZero(self, data):
		try:
			return ''.join(re.search(r'^([A-Z]*)[0 ]*(\d+)$', data).groups())
		except:
			return data
if __name__ == "__main__":
	app = QApplication(sys.argv)
	w = MainWindow()	
	sys.exit(app.exec_())

File: write.php
<?php

error_reporting(E_ERROR | E_PARSE);

$ReaderIp="192.168.1.201";

$port=100;
$timeout = "30";

//Data for writing
$data=$_REQUEST['data'];


/* Create a TCP/IP socket. */
$socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
if ($socket === false) {
    echo "socket_create() failed: reason: " . socket_strerror(socket_last_error()) . "\n";
} else {
    //echo "OK.\n";
}

//echo "Attempting to connect to '$ReaderIp' on port '$port'...";
$result = @socket_connect($socket, $ReaderIp, $port);
if ($result === false) {
    echo "socket_connect() failed.\nReason: ($result) " . socket_strerror(socket_last_error($socket)) . "\n";
} else {
   // echo "OK.\n";
}

//$epc="";
$l=strlen($data);
for($j=0;$j<$l;$j++)
{
	$epc[]= ord(substr($data,$j,1));
}

for($j=0;$j<12-$l;$j++)
{
	$epc[]= 0;
}

//echo $epc."<br>";

$addr=7;
$fail=0;
$succ=0;

for($j=0;$j<6;$j++)
{
$b2=intval($epc[$j*2]);
$b1=intval($epc[$j*2+1]);

//echo $addr+j .",".$b1.",".$b2."<br>";
//echo "Sending Write request... <br>";    10,255,10,137,0,0,0,0,1,3,53,53,246

$cmd = chr(10) . chr(255) . chr(10) . chr(137) .chr(0).chr(0).chr(0).chr(0).chr(1).chr($addr-$j).chr($b1).chr($b2). chr(CheckDigit($addr-$j,$b1,$b2)) ;
//$cmd = chr(10) . chr(255) . chr(05) . chr(129) .chr($addr-$j).chr($b1).chr($b2). chr(246) ;
//echo $cmd . "<br>";

socket_write($socket, $cmd, strlen($cmd));

//echo "Reading response: <br>";

$out = socket_read($socket, 2048);
if(ord(substr($out,3,1))==82) $fail=1;

/*for($i=0;$i<strlen($out);$i++) {	
	echo ord(substr($out,$i,1))."<br>";
	}
	echo "<br>"."<br>";*/
}

//read start
//echo "Sending Read request... <br>";

$cmd = chr(10) . chr(255) . chr(2) . chr(128) . chr(117) ;
socket_write($socket, $cmd, strlen($cmd));

//echo "Reading response: <br>";
$out = socket_read($socket, 2048);
$cnt=ord(substr($out,5,1));

	//echo "Sending get tag data request... <br>";
	$cmd = chr(10) . chr(255) . chr(3) . chr(65).chr(16)  .  chr(163) ;
	socket_write($socket, $cmd, strlen($cmd));

	//echo "Reading response: <br>";
	$out="";
	$out = socket_read($socket, 2048);

	for($j=0;$j<$cnt;$j++)
	{
		$d="";
		for($i=($j*14)+7;$i<($j*14)+19;$i++) {	
			//echo ord(substr($out,$i,1))."<br>";
			if(ord(substr($out,$i,1))!=0) $d= substr($out,$i,1).$d;
		}
		if(ord(substr($d,0,1))==1) $d=substr($d,1);
		if($d==$data1) $succ=1;
			//echo "$d|$data";
	}
	//echo "Reading response: <br>";
	$out="";

//read end
if($succ) echo "Success"; 
else echo "Failed";
die();

function CheckDigit($a, $b, $c)
{
	$i=$a+$b+$c+413;
	
	if($i<255) $i= 256-$i;
	else if ($i<511) $i= 512-$i;
	else if ($i<1023) $i= 1024-$i;
	
	if ($i>255) $i=$i-256;
	
	return $i;
}

?>
